File: DescendantsSamples/ControlExtensions.vb
Public Module ControlExtensions
    <Runtime.CompilerServices.Extension>
    Public Iterator Function Descendants(Of T As Class)(control As Control) As IEnumerable(Of T)
        For Each child As Control In control.Controls
            Dim thisControl As T = TryCast(child, T)
            If thisControl IsNot Nothing Then
                Yield thisControl
            End If

            If child.HasChildren Then
                For Each descendant As T In child.Descendants(Of T)()
                    Yield descendant
                Next
            End If
        Next
    End Function

End Module

File: DescendantsExamples/Extensions/ControlExtensions.cs
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;

namespace DescendantsExamples.Extensions
{
    public static class ControlExtensions
    {
        public static RadioButton RadioButtonChecked(this Control control, bool Checked = true) =>
            control.Descendants<RadioButton>().ToList()
                .FirstOrDefault((radioButton) => radioButton.Checked == Checked);

        public static List<CheckBox> CheckBoxList(this Control control) => 
            control.Descendants<CheckBox>().ToList();
        public static List<ComboBox> ComboBoxList(this Control control) => 
            control.Descendants<ComboBox>().ToList();

        public static IEnumerable<T> Descendants<T>(this Control control) where T : class
        {
            foreach (Control child in control.Controls)
            {
                if (child is T thisControl)
                {
                    yield return (T)thisControl;
                }

                if (child.HasChildren)
                {
                    foreach (T descendant in Descendants<T>(child))
                    {
                        yield return descendant;
                    }
                }
            }
        }
    }
}
